src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -0,0 +1,390 @@
+        self.filesToRecover = []
+
+        logging.info("Component configured to skip container rule creation for data tiers: %s",
+                     self.skipRulesForTiers)
+        logging.info("Component configured to create block rules: %s", self.createBlockRules)
+
+    def setup(self, parameters):
+        """
+        _setup_
+
+        Create DAO Factory and setup some DAO.
+        """
+        myThread = threading.currentThread()
+        daofactory = DAOFactory(package="WMComponent.RucioInjector.Database",
+                                logger=self.logger, dbinterface=myThread.dbi)
+
+        self.getUninjected = daofactory(classname="GetUninjectedFiles")
+        self.getMigrated = daofactory(classname="GetMigratedBlocks")
+
+        self.findDeletableBlocks = daofactory(classname="GetDeletableBlocks")
+        self.markBlocksDeleted = daofactory(classname="MarkBlocksDeleted")
+        self.getUnsubscribed = daofactory(classname="GetUnsubscribedDatasets")
+        self.markSubscribed = daofactory(classname="MarkDatasetSubscribed")
+
+        daofactory = DAOFactory(package="WMComponent.DBS3Buffer",
+                                logger=self.logger, dbinterface=myThread.dbi)
+        self.setStatus = daofactory(classname="DBSBufferFiles.SetPhEDExStatus")
+        self.setBlockClosed = daofactory(classname="SetBlockClosed")
+
+    @timeFunction
+    def algorithm(self, parameters):
+        """
+        _algorithm_
+
+        Poll the database for uninjected files and attempt to inject them into
+        PhEDEx.
+        """
+        if not self.enabled:
+            logging.info("RucioInjector component is disabled in the configuration, exiting.")
+            return
+
+        logging.info("Running Rucio injector poller algorithm...")
+
+        try:
+            # files that failed to get their status updated in dbsbuffer
+            self._updateLFNState(self.filesToRecover, recovery=True)
+
+            # get dbsbuffer_file.in_phedex = 0
+            uninjectedFiles = self.getUninjected.execute()
+
+            # create containers in rucio  (and update local cache)
+            containersAdded = self.insertContainers(uninjectedFiles)
+            if self.containersCache.isCacheExpired():
+                self.containersCache.setCache(containersAdded)
+            else:
+                self.containersCache.addItemToCache(containersAdded)
+
+            # create blocks. Only update the cache once a rule gets created...
+            blocksAdded = self.insertBlocks(uninjectedFiles)
+
+            # now create a Rucio rule for every single block (and update local cache)
+            blocksAdded = self.insertBlockRules(blocksAdded)
+            if self.blocksCache.isCacheExpired():
+                self.blocksCache.setCache(blocksAdded)
+            else:
+                self.blocksCache.addItemToCache(blocksAdded)
+
+            # finally, create file replicas
+            self.insertReplicas(uninjectedFiles)
```

Comment:
@ericvaandering is this how the sync works? by injecting the files *after* creating the rule? I would think its before but am not sure.

Reply:
The Sync makes the rules after, as I recall.

Reply:
Here I first create a rule because I want to make sure there is a rule protecting a block before any replicas get created.
Otherwise, I foresee a day where we could have:
* block is created, then replicas are added
* then the component crashes for the whole weekend (meaning, no rules created)
* what happens to those replicas then?!?

Reply:
I think you should do it as sync: any replicas that have no rules are not possible to be touched by reaper: a replica must have a tombstone set to be reaped and it is only set once all rules are removed. If no rule ever is created, the replica lives on.

Reply:
I can change that. Which brings me to the next question. Should I create a block rule as soon as we have 1 file/replica in the block? Or only when we are about to close the block (or add the last replica)?

Reply:
Well let me rephrase: I think this and the sync code should be consistent. I am not 100% sure which order is better, although at least we have operation experience with rule-after-replica. Things that influence the decision:

- How long it takes rucio to detect changes to DID replicas and move a rule from DONE to REPLICATING
- How a rule behaves when unsatisfiable because no replica exists

Reply:
Let's go with your operation experience then. I'm updating the code.